script to collect upstream build information

sunilangadi2 opened the pull request:
Signed-off-by: Sunil Angadi <sangadi@redhat.com>

# Description
Python script to collect upstream build information
Solves: https://issues.redhat.com/browse/RHCEPHQE-2668

Test Results:
Script returns upstream repo path and container image.
https://3.chacra.ceph.com/repos/ceph/quincy/43e2e60a7559d3f46c9d53f1ca875fd499a1e35e/centos/8/flavors/default/repo
quay.ceph.io/ceph-ci/ceph:43e2e60a7559d3f46c9d53f1ca875fd499a1e35e

vivekanandan-k-rh commented:
LGTM

sunilkumarn417 commented on pipeline/scripts/ci/upstream_build.py:
```diff
@@ -0,0 +1,26 @@
+"""
```
Thanks for adding this script.

Can we make this as stand-alone/independent script which can fetch build information with branch CLI argument.
**Reference :** https://github.com/red-hat-storage/cephci/blob/master/pipeline/scripts/ci/cos_cli.py

The idea is to use this standalone script in upstream pipeline job to fetch build information.

sunilangadi2 replied:
@sunilkumarn417 Thanks for the feedback 
modified the file as per the reference
```(env) sangadi@magna002:~/cephci/cephci/pipeline/scripts/ci$ python3 upstream_cli.py -h
This script helps to provide needed info for upstream specific.

Usage:
  upstream_cli.py rpm branch_name
  upstream_cli.py image branch_name

Example:
    python upstream_cli.py rpm quincy

Options:
    -h --help                       Show this screen
```
Please review.

psathyan commented on pipeline/scripts/ci/upstream_cli.py:
```diff
@@ -0,0 +1,89 @@
+"""
+This script provides the upstream latest rpm repo path and container image.
```
The module does not emit a record... 

- there is no print
- (or) file output giving the intended information.

sunilangadi2 commented:
Thanks, @psathyan for the review, addressed review comments.
@sunilkumarn417 updated the script to get ceph version of upstream branch and the version gets updated here along with 
repo and image http://magna002.ceph.redhat.com/cephci-jenkins/latest-rhceph-container-info/upstream.yaml
Please review the changes.
Test results:
`(env) sangadi@magna002:~/cephci/cephci/pipeline/scripts/ci$ python upstream_cli.py build quincy
2022-05-04 01:40:48,446 - INFO - upstream_cli.py:89 - Upstream repo for the given branch is https://3.chacra.ceph.com/repos/ceph/quincy/f124334e2edf68f8f659f8ca4ed9c4bc6aaeb9bf/centos/8/flavors/default/repo
2022-05-04 01:40:48,447 - INFO - upstream_cli.py:94 - Upstream image for the given branch is quay.ceph.io/ceph-ci/ceph:f124334e2edf68f8f659f8ca4ed9c4bc6aaeb9bf
2022-05-04 01:40:48,588 - INFO - upstream_cli.py:118 - Upstream version for the given branch is ceph-17.2.0-58`

sunilkumarn417 commented on pipeline/scripts/ci/upstream_cli.py:
```diff
@@ -0,0 +1,160 @@
+
+def update_build(branch_name, repo, image, version):
+    # Mount reesi to update changes.
+    commands = [
```
this won't be required, since we execute this script in pipeline agent.

sunilangadi2 replied:
Yes removed it.

sunilkumarn417 commented on pipeline/scripts/ci/upstream_cli.py:
```diff
@@ -0,0 +1,160 @@
+    url = "http://magna002.ceph.redhat.com/cephci-jenkins/latest-rhceph-container-info/upstream.yaml"
+    data = requests.get(url, timeout=30)
+    repo_details = yaml.safe_load(data.text)
+    existing_version = repo_details[branch_name]["ceph-version"]
```
This might fail if it is first time update to the file.

sunilangadi2 replied:
yes handled that condition and raised exception if key i.e branch name not found.

psathyan commented on pipeline/scripts/ci/upstream_cli.py:
```diff
@@ -0,0 +1,160 @@
+        image: upstream branch image
+        version: upstream branch ceph version.
+    """
+    url = "http://magna002.ceph.redhat.com/cephci-jenkins/latest-rhceph-container-info/upstream.yaml"
```
As this script is understand the `pipeline`, we can safely assume that it works only in an CI environment. Hence, we can look at the mount directory for the specific file.

sunilangadi2 replied:
Yes We can do that.

psathyan replied:
You could read the file as the directory is mounted. Unless the intent is to execute the script outside of a pipeline environment.

sunilangadi2 replied:
yes read the file from a mounted directory and removed that URL.

sunilangadi2 commented on pipeline/scripts/ci/upstream_cli.py:
```diff
@@ -0,0 +1,160 @@
+    if exit_status:
+        raise Exception("Please install podman on host")
+
+    cmd = f"sudo podman pull {image}"
```
To confirm that image is valid from the registry, cross validating by pulling the image,
Bcz seen some of the images not able to pull for manifest unknown
`sangadi@sangadi ~]$ sudo podman pull quay.ceph.io/ceph-ci/ceph:ca74598065096e6fcbd8433c8779a2be0c889351
[sudo] password for sangadi: 
Trying to pull quay.ceph.io/ceph-ci/ceph:ca74598065096e6fcbd8433c8779a2be0c889351...
Error: initializing source docker://quay.ceph.io/ceph-ci/ceph:ca74598065096e6fcbd8433c8779a2be0c889351: reading manifest ca74598065096e6fcbd8433c8779a2be0c889351 in quay.ceph.io/ceph-ci/ceph: manifest unknown: manifest unknown`
So that instead moving further we can **Fail** here itself.

Tried pulling the image from a normal user it got failed as below,


> [sangadi@sangadi ~]$ podman pull quay.ceph.io/ceph-ci/ceph:f566756bcbf81a9e17ed272afd5e8555c4ded09b

ERRO[0000] cannot find UID/GID for user sangadi: No subuid ranges found for user "sangadi" in /etc/subuid - check rootless mode in man pages. 
WARN[0000] using rootless single mapping into the namespace. This might break some images. Check /etc/subuid and /etc/subgid for adding sub*ids 
Trying to pull quay.ceph.io/ceph-ci/ceph:f566756bcbf81a9e17ed272afd5e8555c4ded09b...
Getting image source signatures
Copying blob fee18ea417d6 done  
Copying blob 9f45cf1cd9f2 done  
Copying blob f0a2109a2528 done  
Copying blob 32033a8ea5a0 done  
Copying blob bfd1401568a8 done  
Error: writing blob: adding layer with blob "sha256:f0a2109a2528cd1cf5eaa3f5867a6771e4a7c28ae7179d7cc67cbcc2b29f21fc": Error processing tar file(exit status 1): potentially insufficient UIDs or GIDs available in user namespace (requested 0:5 for /usr/bin/write): Check /etc/subuid and /etc/subgid: lchown /usr/bin/write: invalid argument

using **sudo** able to pull the image successfully. hence used as sudo.
@psathyan Please let me know are you Ok with this?

psathyan replied:
Thanks @sunilangadi2 for clarifying.

It would be great if you could add a documentation/comment about this at the top level. This would give us the reason why we need to run this script with root level privileges.

sunilangadi2 commented:
@psathyan Addressed the review comments please review,
also @sunilkumarn417 handled the race condition to avoid multiple write attempts on upstream file for parallel job execution.
Please review.
Test Results:
```
[sangadi@sangadi ci]$ sudo python3 upstream_cli.py build pacific
2022-05-12 15:24:52,695 - INFO - upstream_cli.py:98 - Upstream repo for the given branch is https://1.chacra.ceph.com/repos/ceph/pacific/73636a1b00037ff974bcdc969b009c5ecec626cc/centos/8/flavors/default/repo
2022-05-12 15:24:52,695 - INFO - upstream_cli.py:103 - Upstream image for the given branch is quay.ceph.io/ceph-ci/ceph:73636a1b00037ff974bcdc969b009c5ecec626cc
podman version 3.4.2
2022-05-12 15:24:54,460 - INFO - upstream_cli.py:127 - Upstream version for the given branch is 16.2.7-981
2022-05-12 15:25:00,211 - INFO - upstream_cli.py:59 - Updated build info in /ceph/cephci-jenkins/latest-rhceph-container-info/upstream.yaml

```

psathyan commented on pipeline/scripts/ci/upstream_cli.py:
```diff
@@ -0,0 +1,159 @@
+        raise Exception(f"{branch_name} not found in upstream.yaml")
+    # update only if current version is greater than existing version.
+    if version > existing_version:
+        while True:
```
This might end in a never ending loop. Kindly include a timeout

sunilangadi2 replied:
@psathyan No It's not a never ending loop 
case 1) if it's first time job access the file perform the write and there is Break after closing file so It comes out from while.
case 2) If some job doing some write operation in that file, another job will get exception in opening file as per **fcntl lock** so it will wait for 2 minutes before retrying to access that file once First process finishes it;s write operation second process keeps waiting and retrying.

psathyan replied:
Keeping a timeout would avoid the scenario wherein the lock is held indefinitely by an application. A failure would allow us to be notified and take appropriate steps.

sunilangadi2 replied:
@psathyan Thanks for considering deadlock case handled that with 10 minutes timeout,
Please review.

psathyan reviewed (commented):
LGTM

sunilkumarn417 commented on pipeline/scripts/ci/upstream_cli.py:
```diff
@@ -0,0 +1,166 @@
+    stream = open(file_name, "r")
+    repo_details = yaml.load(stream)
+    # To handle new release which is not updated yet
+    if repo_details[branch_name]:
```
would it make sense to update the release.yaml if release entry is not found? instead of raising exception.

This would give us a way to update dynamically for newer releases.

sunilangadi2 replied:
Yes @sunilkumarn417 it makes sense to update dynamically Instead of raising exceptions I updated the file,
Please review it.

sunilkumarn417 commented on pipeline/scripts/ci/upstream_cli.py:
```diff
@@ -0,0 +1,174 @@
+            "composes": repo,
+            "image": image,
+        }
+        with open(file_name, "w") as yaml_file:
```
needs locking mechanism here as well.

sunilkumarn417 replied:
can we make it common call, once you have all details to be updated.

sunilangadi2 replied:
@sunilkumarn417 It happens only at the first time when new upstream branch is released, So in real time it won't deal with race conditions among jobs, Hence I feel lock not needed there.

sunilangadi2 replied:
@sunilkumarn417 yes handled locking there as well and made it file update as a common call 
Tested both below scenario
1) existing branch update successfully
2) new branch also gets updated in the same format successfully.
Thanks for the review, Please review it also tested both

sunilkumarn417 reviewed (approved):
LGTM.
